q2_types/per_sample_sequences/_format.py

```diff
@@ -160,12 +163,115 @@ def sequences_path_maker(self, sample_id, barcode_id, read_number):
+    http://qiime.org/documentation/file_formats.html#demultiplexed-sequences
+
+    Format details:
+
+    - FASTA file with exactly two lines per record: header and sequence. Each
+      sequence must span exactly one line and cannot be split across multiple
+      lines.
+
+    - The ID in each header must follow the format ``<sample-id>_<seq-id>``.
+      ``<sample-id>`` is the identifier of the sample the sequence belongs to,
+      and ``<seq-id>`` is an identifier for the sequence *within* its sample.
+      In QIIME 1, ``<seq-id>`` is typically an incrementing integer starting
+      from zero, but any non-empty value can be used here, as long as the
+      header IDs remain unique throughout the file. Note: ``<sample-id>`` may
+      contain sample IDs that contain underscores; the rightmost underscore
+      will used to delimit sample and sequence IDs.
+
+    - Descriptions in headers are permitted and ignored.
+
+    - Header IDs must be unique within the file.
+
+    - Each sequence must be DNA and cannot be empty.
+
+    """
+
+    def sniff(self):
+        with self.open() as filehandle:
+            try:
+                self._validate(filehandle, num_records=30)
+            except Exception:
+                return False
+            else:
+                return True
+
+    # The code is structured such that `_validate` can be used to validate as
+    # much of the file as desired. Users may be able to control levels of
+    # validation in the future, and we'll also have the ability to describe
+    # *why* a file is invalid. Sniffers can only offer a boolean response
+    # currently, but the below `Exceptions` could include real error messages
+    # in the future. For now, the `Exceptions` are only used to give a boolean
+    # response to the sniffer.
+    def _validate(self, filehandle, *, num_records):
+        ids = set()
+        for (header, seq), _ in zip(itertools.zip_longest(*[filehandle] * 2),
+                                    range(num_records)):
+            if header is None or seq is None:
+                # Not exactly two lines per record.
+                raise Exception()
+
+            header = header.rstrip('\n')
+            seq = seq.rstrip('\n')
+
+            id = self._parse_id(header)
+            if id in ids:
+                # Duplicate header ID.
+                raise Exception()
+
+            self._validate_id(id)
+            self._validate_seq(seq)
+
+            ids.add(id)
+
+        if not ids:
+            # File was empty.
+            raise Exception()
+
+    def _parse_id(self, header):
+        if not header.startswith('>'):
+            raise Exception()
+        header = header[1:]
+
+        id = ''
+        if header and not header[0].isspace():
+            id = header.split(maxsplit=1)[0]
+        return id
+
+    def _validate_id(self, id):
+        pieces = id.rsplit('_', maxsplit=1)
+        if len(pieces) != 2 or not pieces[0] or not pieces[1]:
```

Comment:
This is silly, but I think `not all(pieces)` would work here over indexing.

Reply:
Thanks, that is much cleaner!